File: local-data-persistent/LatihanRepository (finish)/java/app/src/main/java/com/dicoding/newsapp/data/local/room/NewsDatabase.java
package com.dicoding.newsapp.data.local.room;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.dicoding.newsapp.data.local.entity.NewsEntity;

@Database(entities = {NewsEntity.class},
        version = 1,
        exportSchema = false)
public abstract class NewsDatabase extends RoomDatabase {
    public abstract NewsDao newsDao();

    private static volatile NewsDatabase INSTANCE;

    public static NewsDatabase getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (NewsDatabase.class) {
                INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                        NewsDatabase.class, "News.db")
                        .build();
            }
        }
        return INSTANCE;
    }
}

File: local-data-persistent/LatihanRepository (coroutines)/app/src/main/java/com/dicoding/newsapp/data/local/room/NewsDatabase.kt
package com.dicoding.newsapp.data.local.room

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.dicoding.newsapp.data.local.entity.NewsEntity

@Database(entities = [NewsEntity::class], version = 1, exportSchema = false)
abstract class NewsDatabase : RoomDatabase() {
    abstract fun newsDao(): NewsDao

    companion object {
        @Volatile
        private var instance: NewsDatabase? = null
        fun getInstance(context: Context): NewsDatabase =
            instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(
                    context.applicationContext,
                    NewsDatabase::class.java, "News.db"
                ).build()
            }
    }
}